File: backend/schemas/questions.py
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field


AnswerType = Literal["text", "single_select", "multi_select", "number", "boolean"]


class QuestionCreateRequest(BaseModel):
    question_id: str = Field(min_length=2, max_length=100)
    question_text: str = Field(min_length=3)
    category: Optional[str] = None
    answer_type: AnswerType
    options: list[str] = Field(default_factory=list)
    is_required: bool = False
    is_active: bool = True
    version: int = 1


class QuestionUpdateRequest(BaseModel):
    question_text: Optional[str] = Field(default=None, min_length=3)
    category: Optional[str] = None
    answer_type: Optional[AnswerType] = None
    options: Optional[list[str]] = None
    is_required: Optional[bool] = None
    is_active: Optional[bool] = None
    version: Optional[int] = None


class PendingSetsResponse(BaseModel):
    """Which question sets this user has a set for and has answered none of."""

    party: bool = False
    position: bool = False


class QuestionResponse(BaseModel):
    id: str
    question_id: str
    question_text: str
    category: Optional[str]
    answer_type: AnswerType
    options: list[str]
    is_required: bool
    is_active: bool
    version: int
    created_at: datetime
    updated_at: datetime
    # Position and party questions only, absent on profile questions. Which
    # party the question belongs to, which position group (position questions
    # only), its place in that set, and its Hindi. options_hi runs parallel to
    # options: the English option is what gets saved, the Hindi at the same
    # index is what gets shown.
    party: Optional[str] = None
    position_group: Optional[str] = None
    display_order: Optional[int] = None
    question_text_hi: Optional[str] = None
    options_hi: list[str] = Field(default_factory=list)
    # Separate from is_required, which the batch save enforces across every
    # active question at once and so cannot be used for a question only half
    # the users are ever shown. This one only marks the question in the UI.
    is_compulsory: bool = False
    # Party questions only. The option to show selected before the user has
    # chosen one, and the same option the prompt falls back to, so the screen
    # and the generated post never disagree.
    default_option: Optional[str] = None

File: frontend/src/api/questions.js
import client from './client';

// GET /questions — fetch active questions ordered by creation
// Returns [{ question_id, question_text, options, answer_type, ... }]
// Position questions are not in this list; use getPositionQuestions.
export async function getQuestions(limit = 7) {
  const { data } = await client.get('/questions/', { params: { limit } });
  return data;
}

// GET /questions/position — the five preference questions written for one
// party and one position group, in order. Empty for a party or group that has
// no set yet, which is an ordinary answer rather than an error.
//
// `party` is the party name exactly as stored on the user; the backend decides
// which set it maps to. Each question carries options_hi parallel to options:
// save the English option, show the Hindi at the same index.
export async function getPositionQuestions(party, group) {
  const { data } = await client.get('/questions/position', { params: { party, group } });
  return Array.isArray(data) ? data : [];
}

// GET /questions/party — the ten questions about the party itself, in order.
// These ask what the writer wants said about their party rather than how
// someone at their level writes, so they depend on the party alone and not on
// the position. Empty for a party with no set, which is an ordinary answer.
//
// Questions carry is_compulsory. It is deliberately not is_required: both
// parties' sets are active at once, so the backend cannot enforce it across
// every user. The screens that ask the question enforce it instead.
export async function getPartyQuestions(party) {
  const { data } = await client.get('/questions/party', { params: { party } });
  return Array.isArray(data) ? data : [];
}

// GET /questions/pending — { party, position }, each true when the signed-in
// user has that set and has answered none of it.
//
// "None of it", not "some gap in it": someone who answered two of ten has seen
// the questionnaire and stopped, and the rest fall back to their defaults.
//
// The party set is only ever pending for an account older than the questions,
// which is what keeps a fresh sign-up at five questions instead of fifteen.
export async function getPendingQuestionSets() {
  const { data } = await client.get('/questions/pending');
  return { party: Boolean(data?.party), position: Boolean(data?.position) };
}
